internal/server/server.go

```diff
@@ -32,32 +31,62 @@ func Run(opts ...Option) error {
 		cfg.Logger = hclog.L()
 	}
 
-	// Setup our run group since we're going to be starting multiple
-	// goroutines for all the servers that we want to live/die as a group.
-	var group run.Group
-
-	// We first add an actor that just returns when the context ends. This
-	// will trigger the rest of the group to end since a group will not exit
-	// until any of its actors exit.
-	ctx, cancelCtx := context.WithCancel(cfg.Context)
-	cfg.Context = ctx
-	group.Add(func() error {
-		<-ctx.Done()
-		return ctx.Err()
-	}, func(error) { cancelCtx() })
+	log := cfg.Logger.Named("server")
 
-	// Setup our gRPC server.
-	if err := grpcInit(&group, &cfg); err != nil {
+	grpcServer, err := newGrpcServer(&cfg)
+	if err != nil {
 		return err
 	}
 
-	// Setup our HTTP server.
-	if err := httpInit(&group, &cfg); err != nil {
-		return err
+	wg := sync.WaitGroup{}
+
+	gprcErrs := make(chan error, 1)
+	wg.Add(1)
+	go func() {
+		err := grpcServer.start()
+		gprcErrs <- err
+		log.Debug("gRPC server has exited", "err", err)
+		wg.Done()
+	}()
+
+	httpErrs := make(chan error, 1)
+	httpEnabled := cfg.HTTPListener != nil
+	var httpServer *httpServer
+	if httpEnabled {
+		wg.Add(1)
+		httpServer = newHttpServer(grpcServer.server, &cfg)
+		go func() {
+			err := httpServer.start()
+			httpErrs <- err
+			log.Debug("HTTP server has exited", "err", err)
+			wg.Done()
+		}()
 	}
 
-	// Run!
-	return group.Run()
+	ctx, cancel := context.WithCancel(cfg.Context)
+	defer cancel()
+
+	defer wg.Wait() // Wait for server run processes to exit before returning
+	select {
+	case err := <-gprcErrs:
+		// If the GRPC server errored, we can assume it's closed and shut down the http server if necessary
+		if httpEnabled {
+			httpServer.close()
+		}
+		return fmt.Errorf("failed running the grpc server: %w", err)
```

Comment:
We shouldn't re-wrap the error returned from starting the server here, especially if the original error contains any `status` `codes` from grpc, We lose them if we do this I think and can't check the kind of error (i.e. we can't check later that `status.Code(err) == codes.gRPC_Status_Code`). Instead we can log above the return that we failed to run the grpc server and return the error as given?

Same comment below for failing to run the http server.

Reply:
👍 Logging the error too seems like a good idea here, because it looks like the error is discarded upstream.